src/main/java/net/minestom/server/instance/DynamicChunk.java

```diff
@@ -96,49 +102,47 @@ public void setBlock(int x, int y, int z, @NotNull Block block,
         int sectionRelativeX = globalToSectionRelative(x);
         int sectionRelativeZ = globalToSectionRelative(z);
 
-        section.blockPalette().set(
-                sectionRelativeX,
-                globalToSectionRelative(y),
-                sectionRelativeZ,
-                block.stateId()
-        );
-
         final int index = CoordConversion.chunkBlockIndex(x, y, z);
         // Handler
         final BlockHandler handler = block.handler();
-        final Block lastCachedBlock;
-        if (handler != null || block.hasNbt() || block.registry().isBlockEntity()) {
-            lastCachedBlock = this.entries.put(index, block);
-        } else {
-            lastCachedBlock = this.entries.remove(index);
-        }
+        final boolean shouldCache = handler != null || block.hasNbt() || block.registry().isBlockEntity();
+        final Block lastCachedBlock = shouldCache ? this.entries.get(index) : null;
```

Comment:
this should just be a remove always, we are going to remove the old block regardless of whether the current block needs to be stored.